### PR TITLE
[rest] PersistenceResource: Support transforming states to display states

### LINKED ISSUE #4697: [charts] add option to show state desciptions instead of value
## The problem
i have a diagram that shows me the numerical operating state of a device. i can see in the diagram that the device changes from state 0 to state 5 at time x, for example. the operating states are defined in the item itself:
```
  "stateDescription": {
    "pattern": "%.0f",
    "readOnly": true,
    "options": [
      {
        "value": "0",
        "label": "Heating"
      },
      {
        "value": "1",
        "label": "Hot water"
      },
      {
        "value": "2",
        "label": "Swimming pool / Photovoltaics"
      },
      {
        "value": "3",
        "label": "Utility lock"
      },
      {
        "value": "4",
        "label": "Defrost"
      },
      {
        "value": "5",
        "label": "No requirement"
      },
      {
        "value": "6",
        "label": "Heating ext. energy source"
      },
      {
        "value": "7",
        "label": "Cooling mode"
      }
    ]
  }

```
it would be very useful, if the charts show the appropriate label (as a legend or above the measurement line...) instead of the values. it could be an option in the config menü i.e..

## Your suggestion
Add an option to config menue of a chart. I.e.: "use StateDescription instead of values"

## Your environment
OH3.4.2

### PR DESCRIPTION
**Pull request by florian-h05**
Resolves #4697.

This adds a new query param to the GET `/rest/persistence/items/{itemname}` endpoint that enables transformation of the persisted states according to the state description of the Item, so that display states can be shown in charts.
It also adds a unit field to the response DTO.
Extracting the display state calculation into its own utility class is also a required step for adding a WS with similar functionality as the itemstates SSE endpoint.

### CODE REVIEW

**Comment by mherwege** [comment]
As you create a utility class for this, I am wondering if it should be considered here as well to keep everything aligned: https://github.com/openhab/openhab-core/blob/1e33058bb7edac30f3e6babec6ad12d27c7aa746/bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/items/ItemUIRegistryImpl.java#L343

**Comment by florian-h05** [comment]
> As you create a utility class for this, I am wondering if it should be considered here as well to keep everything aligned:

That's a bit difficult since this class does a few things more and has some override behaviour, but I'll try.

**Comment by florian-h05** [comment]
@mherwege Done. With some help from Gemini CLI I got all ItemUIRegistryImplTest tests green after my first, radical refactoring attempt.

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java [hunk 1]
```diff
@@ -500,12 +509,36 @@ private Response getItemHistoryDTO(@Nullable String serviceId, String itemName,
...
+                        targetUnit = UnitUtils.parseUnit(stateDescription.getPattern());
+                    }
+                }
+                if (targetUnit == null) {
+                    targetUnit = numberItem.getUnit();
+                }
+            }
+        } catch (ItemNotFoundException e) {
+            logger.debug("Failed to retrieve Item '{}' from registry", itemName);
+            return null;
```
[the remark this answers is not in the document]
Comment (florian-h05) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java]: Copilot suggested that change itself in the last review. I am a bit unsure what behaviour is intended here. Normally, I'd say when an Item does not exist, one doesn't need to read its persisted data, but technically this is possible right now.
@mherwege WDYT?

[the remark this answers is not in the document]
Comment (mherwege) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java]: I agree with copilot on this. I also think we should still keep the possibility to query all data in persistence through the REST API, even if there is no longer an item defined for it. It can help with persistence maintenance.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java [hunk 2]
```diff
@@ -33,14 +34,16 @@ public class ItemHistoryDTO {
     public String name;
     public String datapoints;
 
+    @Schema(description = "The unit for the returned values (if available)")
+    public @Nullable String unit;
```
Comment (mherwege) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java:38]: I am just wondering why you need the unit here. The HistoryDataBean includes a State field that carries the unit. When querying persistence for a QuantityType Item, it should return the unit. As you want to adjust this to the displayState unit, shouldn't you change it in each of the returned records?
Be aware that some persistence services may be able to store a unit with values and the unit returned may even be different for each value. This is somewhat theoretical, but should be possible. Only when the persistence service does not store units, the item unit is assumed.
EDIT: I see, current logic is stripping the unit returned from persistence. So you need a field here. I see a problem with just stripping the unit and assuming it is the item unit though.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java [hunk 3]
```diff
@@ -60,6 +63,19 @@ public void addData(long time, State state) {
...
     }
 
+    /**
+     * Add a new record to the data history.
+     *
+     * @param time the time of the record
+     * @param state the state at this time
+     */
+    public void addData(long time, String state) {
+        HistoryDataBean newVal = new HistoryDataBean();
```
Comment (mherwege) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java:73]: If you convert the state if it is a QuantityType to the expected unit in the first place, you don't need to introduce string states here.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java
Comment (mherwege) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java:56]: I actually think this is an assumption that should not be made. If you have a persistence service that stores units, it may not be true.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java (hunk 3 above)
Comment (florian-h05) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java:73]: But I generally support transforming to the displayState, and that logic only returns a string. Also, when converting to the unit from the display state for QuantityTypes, I still can't enforce the number of decimals as defined in the state description pattern when passign a QuantityType here.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java
Comment (florian-h05) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java:56]: That assumption was there before I touched the code, I would prefer to keep this out of scope here.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java (hunk 2 above)
Comment (florian-h05) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java:38]: In reality, the bean carries a value without a unit (as it is stripped above, as you also commented on). As the unit is not expected (I assume this for now) to change between the persisted points, it is more efficient to just add the unit top-level to the DTO compared to adding it to every single value.

**Comment by florian-h05** [comment]
My work on this is finished (except for review of course). I have completed the refactoring of ItemUIRegistryImpl.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java (hunk 3 above)
Comment (mherwege) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/dto/ItemHistoryDTO.java:73]: OK.

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java [hunk 4]
```diff
@@ -604,6 +644,31 @@ private Response getItemHistoryDTO(@Nullable String serviceId, String itemName,
...
+                    logger.warn(
+                            "Cannot convert state '{}' to unit '{}' for item '{}', excluding this state from the response",
+                            state, targetUnit, item.getName());
+                }
+            } else {
+                String displayStateStr = getDisplayState(item, state, locale);
+                dto.addData(timestamp, displayStateStr != null ? displayStateStr : state.toString());
+            }
+        } else {
+            dto.addData(timestamp, state);
```
Comment (mherwege) [inline bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java:676]: I think you should make sure the state is in the item unit here, and if not (and there is an item in the item registry) convert to it. Persistence services don't always return in item unit.

Comment (florian-h05) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java:676]: Done.

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (florian-h05) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java]: Changed.

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java [hunk 5]
```diff
@@ -500,12 +513,54 @@ private Response getItemHistoryDTO(@Nullable String serviceId, String itemName,
...
+        @Nullable
+        Unit<?> targetUnit = null;
+        @Nullable
+        String pattern = null;
+        List<StateOption> options = List.of();
+        Item item = null;
+        try {
+            item = itemRegistry.getItem(itemName);
+            if (displayState) {
+                StateDescription stateDescription = item.getStateDescription(locale);
```
Comment (mherwege) [inline bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java]: You can avoid calling this multiple times by moving it one line up. It is also called in line 537.

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java [hunk 6]
```diff
@@ -500,12 +513,54 @@ private Response getItemHistoryDTO(@Nullable String serviceId, String itemName,
...
+                    pattern = stateDescription.getPattern();
+                    options = stateDescription.getOptions();
+                }
+            }
+            if (item instanceof NumberItem numberItem) {
+                if (displayState) {
+                    if (pattern != null) {
+                        targetUnit = UnitUtils.parseUnit(pattern);
+                    }
+                    if (targetUnit == null) {
```
Comment (mherwege) [inline bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java]: Why can this happen? The pattern should already be set in line 527 and the next lines do the same thing.

**Review by mherwege — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java (hunk 6 above)
Comment (florian-h05) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java]: Corrected

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java (hunk 5 above)
Comment (florian-h05) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java]: Changed

**Comment by florian-h05** [comment]
@mherwege Ready for your review again - I also added a few additional test cases.

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java [hunk 7]
```diff
@@ -521,24 +521,16 @@ private Response getItemHistoryDTO(@Nullable String serviceId, String itemName,
...
-                    pattern = stateDescription.getPattern();
-                    options = stateDescription.getOptions();
-                }
+            StateDescription stateDescription = item.getStateDescription(locale);
+            if (displayState && stateDescription != null) {
+                pattern = stateDescription.getPattern();
+                options = stateDescription.getOptions();
             }
             if (item instanceof NumberItem numberItem) {
                 if (displayState) {
```
Comment (mherwege) [inline bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java]:
```
if (displayState && pattern != null) {
```

**Review by mherwege — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java (hunk 7 above)
Comment (florian-h05) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java]: Changed

**Review by mherwege — APPROVED** [review]
LGTM

**Comment by lolodomo** [comment]
That PR changed a very sensible and complex part of sitemap UI, I hope it broke nothing.

**Comment by florian-h05** [comment]
There are tons of unit tests for sitemap getLabel(Widget w) that are all still green, so I'm confident it broke none of the common use cases. Edge cases that aren't covered in the tests might have changed behaviour.

**Comment by mherwege** [comment]
I strongly believe we should strive for consistency and remove code duplication as much as possible. That's why I asked @florian-h05 to consider using the new utility in this part of the code as well. I know it is a complex part, but as pointed out, there are a lot of tests that still pass, and I believe the code is easier to maintain because of this change. It also helps keeping consistency for the user in the interpretation of the formats, as there should be less surprises about something applying in one place and not in another with the same underlying configuration (state description). I can't be 100% sure, but I think this should be fine. If you find something, I am happy to look into it.
